apps/api/v2/src/ee/bookings/2024-08-13/guards/booking-pbac.guard.ts

```diff
@@ -0,0 +1,205 @@
+      throw new ForbiddenException(
+        `BookingPbacGuard - user with id=${user.id} does not have access to booking with uid=${bookingUid}`
+      );
+    }
+
+    const hasRequiredPermissions = await this.checkUserHasRequiredPermissions(
+      user.id,
+      teamId,
+      requiredPermissions
+    );
+
+    if (!hasRequiredPermissions) {
+      await this.setCacheRequiredPermissions(user.id, bookingUid, requiredPermissions, false);
+      throw new ForbiddenException(
+        `BookingPbacGuard - user with id=${user.id} does not have the required permissions=${requiredPermissions.join(",")} for team with id=${teamId}`
+      );
+    }
+
+    await this.setCacheRequiredPermissions(user.id, bookingUid, requiredPermissions, true);
+    request.pbacAuthorizedRequest = true;
+    return true;
+  }
+
+  async hasPbacEnabled(teamId: number) {
+    const cachedHasPbacEnabled = await this.getCachePbacEnabled(teamId);
+
+    if (cachedHasPbacEnabled !== null) {
+      return cachedHasPbacEnabled;
+    }
+
+    const pbacFeatureFlag = "pbac";
+    const featuresRepository = new FeaturesRepository(this.prismaReadService.prisma);
+    const hasPbacEnabled = await featuresRepository.checkIfTeamHasFeature(teamId, pbacFeatureFlag);
+
+    if (hasPbacEnabled) {
+      await this.setCachePbacEnabled(teamId, hasPbacEnabled);
+    }
+
+    return hasPbacEnabled;
+  }
+
+  async checkUserHasRequiredPermissions(
+    userId: number,
+    teamId: number,
+    requiredPermissions: PermissionString[]
+  ) {
+    const permissionCheckService = new PermissionCheckService();
+    const hasRequiredPermissions = await permissionCheckService.checkPermissions({
+      userId,
+      teamId,
+      permissions: requiredPermissions,
+      fallbackRoles: [],
+    });
+
+    return hasRequiredPermissions;
+  }
+
+  private async getCacheRequiredPermissions(
+    userId: number,
+    bookingUid: string,
+    requiredPermissions: PermissionString[]
+  ): Promise<boolean | null> {
+    return this.redisService.get<boolean>(
+      REDIS_BOOKING_REQUIRED_PERMISSIONS_CACHE_KEY(userId, bookingUid, requiredPermissions)
+    );
+  }
+
+  private async setCacheRequiredPermissions(
+    userId: number,
+    bookingUid: string,
+    requiredPermissions: PermissionString[],
+    hasRequired: boolean
+  ): Promise<void> {
+    await this.redisService.set<boolean>(
+      REDIS_BOOKING_REQUIRED_PERMISSIONS_CACHE_KEY(userId, bookingUid, requiredPermissions),
+      hasRequired,
+      { ttl: 300_000 }
+    );
+  }
+
+  private async getCachePbacEnabled(teamId: number): Promise<boolean | null> {
+    const cachedResult = await this.redisService.get<boolean>(REDIS_BOOKING_PBAC_CACHE_KEY(teamId));
+    return cachedResult;
+  }
+
+  private async setCachePbacEnabled(teamId: number, pbacEnabled: boolean) {
+    await this.redisService.set<boolean>(REDIS_BOOKING_PBAC_CACHE_KEY(teamId), pbacEnabled, {
+      ttl: 300_000,
```

Comment:
It looks like the decision is solely based on teamId and thus we could use the same cache key that is agnostic of _teamId_ as in pbac.guard.ts. Infact we should try to reuse the same methods.

https://github.com/calcom/cal.com/blob/feat/add-sesssion-endpoint/apps/api/v2/src/modules/auth/guards/pbac/pbac.guard.ts#L19